lib/mongo/auth/aws/credentials_retriever.rb

```diff
@@ -0,0 +1,181 @@
+      class CredentialsRetriever
+
+        # Timeout for metadata operations, in seconds.
+        #
+        # The auth spec suggests a 10 second timeout but this seems
+        # excessively long given that the endpoint is essentially local.
+        METADATA_TIMEOUT = 5
+
+        def initialize(user = nil)
+          @user = user
+        end
+
+        # @return [ Auth::User | nil ] The user object, if one was provided.
+        attr_reader :user
+
+        # Retrieves a valid set of credentials, if possible, or raises
+        # Auth::InvalidConfiguration.
+        #
+        # @return [ Auth::Aws::Credentials ] A valid set of credentials.
+        #
+        # @raise Auth::InvalidConfiguration if credentials could not be
+        #   retrieved for any reason, or if a source contains an invalid set
+        #   of credentials.
+        def credentials
+          if user
+            credentials = Credentials.new(
+              user.name,
+              user.password,
+              user.auth_mech_properties['aws_session_token'],
+            )
+
+            if credentials_valid?(credentials, 'Mongo::Client URI or Ruby options')
+              return credentials
+            end
+          end
+
+          credentials = Credentials.new(
+            ENV['AWS_ACCESS_KEY_ID'],
+            ENV['AWS_SECRET_ACCESS_KEY'],
+            ENV['AWS_SESSION_TOKEN'],
+          )
+
+          if credentials_valid?(credentials, 'environment variables')
+            return credentials
+          end
+
+          credentials = ec2_metadata_credentials
+
+          if credentials && credentials_valid?(credentials, 'EC2 instance metadata')
+            return credentials
+          end
+
+          raise Auth::InvalidConfiguration,
+            "Could not locate AWS credentials (checked Client URI and Ruby options, environment variables, EC2 metadata)"
+        end
+
+        private
+
+        # Returns credentials from the EC2 metadata endpoint. The credentials
+        # could be empty, partial or invalid.
+        #
+        # @return [ Auth::Aws::Credentials | nil ] A set of credentials, or nil
+        #   if retrieval failed.
+        def ec2_metadata_credentials
+          http = Net::HTTP.new('169.254.169.254')
+          req = Net::HTTP::Put.new('/latest/api/token',
+            # The TTL is required in order to obtain the metadata token.
+            {'x-aws-ec2-metadata-token-ttl-seconds' => '30'})
+          resp = ::Timeout.timeout(METADATA_TIMEOUT) do
+            http.request(req)
+          end
+          if resp.code != '200'
+            return nil
+          end
+          metadata_token = resp.body
+          resp = ::Timeout.timeout(METADATA_TIMEOUT) do
+            http_get(http, '/latest/meta-data/iam/security-credentials', metadata_token)
+          end
+          if resp.code != '200'
+            return nil
+          end
+          role_name = resp.body
+          escaped_role_name = CGI.escape(role_name).gsub('+', '%20')
```

Comment:
Should this use `URI::encode`?

Reply:
`URI::encode` is deprecated and the Ruby driver no longer uses it.

Reply:
Oh sorry. I missed the `CGI.escape` and thought the code was _only_ escaping `+`. Please ignore.